pkg/docker/docker_test.go

```diff
@@ -24,201 +30,336 @@ func TestContainerName(t *testing.T) {
 }
 
 func getDocker(client Client) *stiDocker {
+	//k8s has its own fake docker client mechanism
+	k8sDocker := dockertools.ConnectToDockerOrDie("fake://")
 	return &stiDocker{
-		client:   client,
-		pullAuth: docker.AuthConfiguration{},
+		kubeDockerClient: k8sDocker,
+		client:           client,
+		pullAuth:         dockertypes.AuthConfig{},
 	}
 }
 
+//NOTE, neither k8s kube client nor engine api make their error types public, so we can access / instantiate them like we did for the analogous
+// ones in go-dockerclient.
+// Hence, we test with errors created in these tests and set on the fake client  vs. those actually defined in k8s/engine-api.
+// Also, the fake client does save the parameters passed in; but instead, you can confirm that it was called via AssertCalls
+
 func TestIsImageInLocalRegistry(t *testing.T) {
 	type testDef struct {
-		imageName         string
-		docker            test.FakeDockerClient
-		expectedImageName string
-		expectedResult    bool
-		expectedError     error
+		imageName      string
+		docker         test.FakeDockerClient
+		expectedResult bool
+		expectedError  string
 	}
-	otherError := fmt.Errorf("Other")
 	tests := map[string]testDef{
-		"ImageFound":    {"a_test_image", test.FakeDockerClient{Image: &docker.Image{}}, "a_test_image:latest", true, nil},
-		"ImageNotFound": {"a_test_image:sometag", test.FakeDockerClient{InspectImageErr: []error{docker.ErrNoSuchImage}}, "a_test_image:sometag", false, nil},
-		"ErrorOccurred": {"a_test_image", test.FakeDockerClient{InspectImageErr: []error{otherError}}, "a_test_image:latest", false, otherError},
+		"ImageFound":    {"a_test_image", test.FakeDockerClient{}, true, ""},
+		"ImageNotFound": {"a_test_image:sometag", test.FakeDockerClient{}, false, "unable to get metadata for a_test_image:sometag"},
 	}
 
 	for test, def := range tests {
 		dh := getDocker(&def.docker)
+		fake := dh.kubeDockerClient.(*dockertools.FakeDockerClient)
+		if def.expectedResult {
+			fake.Image = &dockertypes.ImageInspect{ID: def.imageName}
+		}
+
 		result, err := dh.IsImageInLocalRegistry(def.imageName)
+
+		if e := fake.AssertCalls([]string{"inspect_image"}); e != nil {
+			t.Errorf("%+v", e)
+		}
+
 		if result != def.expectedResult {
 			t.Errorf("Test - %s: Expected result: %v. Got: %v", test, def.expectedResult, result)
 		}
-		if err != def.expectedError {
-			t.Errorf("Test - %s: Expected error: %v. Got: %v", test, def.expectedError, err)
-		}
-		if def.docker.InspectImageName[0] != def.expectedImageName {
-			t.Errorf("Docker inspect called with unexpected image name: %s",
-				def.docker.InspectImageName)
+		if err != nil && len(def.expectedError) > 0 && !strings.Contains(err.Error(), def.expectedError) {
+			t.Errorf("Test - %s: Expected error: Got: %+v", test, err)
 		}
 	}
 }
-
 func TestCheckAndPullImage(t *testing.T) {
+	// in addition to the deltas mentioned up top, kube fake client can't distinguish between exists and successful pull, and
```

Comment:
On Thu, Aug 4, 2016 at 5:34 PM, Gabe Montero <gmontero@redhat.com> wrote:

> All the wrappers of errors at the different layers and private error types
> is making it difficult, but I've
> only spent a little bit of time... I'll circle back to this, spend some
> more time, and report back.
>

OK, so I've got a *form* of err image not found test ... i create an error
with "no such image" in the text and inject it into the fake client, and
then
confirm that the ultimate CheckAndPullImage call does not return an error.

>
> On Thu, Aug 4, 2016 at 5:25 PM, Ben Parees <notifications@github.com>
> wrote:
>
>> In pkg/docker/docker_test.go
>> <https://github.com/openshift/source-to-image/pull/554#discussion_r73605731>
>> :
>>
>> >  func TestCheckAndPullImage(t *testing.T) {
>> > +	// in addition to the deltas mentioned up top, kube fake client can't distinguish between exists and successful pull, and
>>
>> ok so doesn't that solve the problem of how to force an ErrImageNotFound
>> error?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/openshift/source-to-image/pull/554/files/63881032417630cb32a9bf66a2b033fee54d2aff..068c1c2bd2a589952a1bd740f976bca1c1fd8657#r73605731>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ADbadPxz3CkTOwm2yrgyPjX7JtYnFd1sks5qcljKgaJpZM4JUVnn>
>> .
>>
>
>